### PR TITLE
2924/fix hash character in user endpoint

### LINKED ISSUE #2924: cannot delete custom user attribute due to "#" in username
### Top-level intent

What did you try to achieve?
tried deleting a custom user attribute that has been set earlier to a specific user containing a "#" in the username (e.g. testuser#test)

### Steps to reproduce

1. assign any custom user attribute (e.g. testattribute) with any value (e.g. 1) to a user containing a "#" sign in the username (e.g. testuser#test)
2. try to remove the assigned custom user attribute "testattribute" from the user

### Expected outcome

What do you think, how the system should have worked?
the assigned user attribute "testattribute" should've been removed from the specific user

### Actual outcome

What did actually happen?
custom user attribute "testattribute" didn't got deleted
error in the GUI: "Failed to get a valid JSON response from the privacyIDEA server."
developer console log in the browser shows a HTTP 404 with the following URL/URI: https://192.168.xxx.yyy/user/attribute/testattribute/testuser <-- notice the username got "cut" after the "#" sign

### Configuration

* **privacyIDEA version**: 3.6
* **Installation method**: from Ubuntu packages
* **Python version**: 3.6.9
* **Operating system**: Ubuntu 18.04 LTS
* **Webserver**: Apache/2.4.29 (Ubuntu)
* **Token database**: MySQL

### Log file

**Set PI_LOGLEVEL = logging.DEBUG in pi.cfg and take a look at the privacyidea.log!**
**If appropriate, attach the log file or paste relevant portions.**
```
[Tue Nov 23 17:31:40.961351 2021] [wsgi:error] [pid 32651] [remote 192.168.xxx.yyy:59891] DEBUG:privacyidea.api.before_after:Begin handling of request '/user/attribute/testattribute/testuser?'
[Tue Nov 23 17:31:40.962493 2021] [wsgi:error] [pid 32651] [remote 192.168.xxx.yyy:59891] DEBUG:privacyidea.api.before_after:End handling of request '/user/attribute/testattribute/testuser?'
```

```
root@ubuntu:/etc/apache2/sites-enabled# http --verify no DELETE https://localhost/user/attribute/testattribute/testuser#test Authorization:...
HTTP/1.1 404 NOT FOUND
Connection: Keep-Alive
Content-Length: 232
Content-Type: text/html; charset=utf-8
Date: Tue, 23 Nov 2021 16:14:05 GMT
Keep-Alive: timeout=5, max=100
Server: Apache

<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 3.2 Final//EN">
<title>404 Not Found</title>
<h1>Not Found</h1>
<p>The requested URL was not found on the server. If you entered the URL manually please check your spelling and try again.</p>
```

When I try executing the same call with a user w/o a "#" in the username, I'm getting the expected result (both API and in GUI):
```
root@ubuntu:/etc/apache2/sites-enabled# http --verify no DELETE https://localhost/user/attribute/testattribute/testuser2 Authorization:...
HTTP/1.1 200 OK
Cache-Control: no-cache
Connection: Keep-Alive
Content-Length: 693
Content-Type: application/json
Date: Tue, 23 Nov 2021 16:15:10 GMT
Keep-Alive: timeout=5, max=100
Server: Apache

{
    "id": 1,
    "jsonrpc": "2.0",
    "result": {
        "status": true,
        "value": 1
    },
    "signature": "rsa_sha256_pss:...",
    "time": 1637684110.8554983,
    "version": "privacyIDEA 3.6",
    "versionnumber": "3.6"
}
```

### PR DESCRIPTION
**Pull request by 1-am-r00t**
fixes #2924 (add url encoding in user endpoint)

### CODE REVIEW

File: privacyidea/static/components/user/factories/user.js [hunk 1]
```diff
@@ -87,7 +87,7 @@ myApp.factory("UserFactory", ['AuthFactory', '$http', '$state', '$rootScope',
                     function(error) {AuthFactory.authError(error.data)});
             },
             deleteCustomAttribute: function(username, realmname, key, callback) {
-                $http.delete(userUrl + "/attribute/" + key + "/" + username + "/" + realmname,
+                $http.delete(userUrl + "/attribute/" + encodeURIComponent(key) + "/" + encodeURIComponent(username) + "/" + encodeURIComponent(realmname),
```
Comment (plettich) [inline privacyidea/static/components/user/factories/user.js:90]: Same as above for the `realmname`.

**Review by plettich — CHANGES_REQUESTED** [review]

**Comment by plettich** [comment]
Hi @1-am-r00t
thanks for the PR.
I think these are the only places where the username is part of the URL.

File: privacyidea/static/components/user/factories/user.js (hunk 1 above)
Comment (1-am-r00t) [reply privacyidea/static/components/user/factories/user.js]: proposed changes implemented in [bfc9af4](https://github.com/privacyidea/privacyidea/pull/3306/commits/bfc9af4ac9a8d357b4fa1e74452474bf89a60d74)

File: privacyidea/static/components/user/factories/user.js [hunk 2]
```diff
@@ -56,7 +56,7 @@ myApp.factory("UserFactory", ['AuthFactory', '$http', '$state', '$rootScope',
                             function(error) { AuthFactory.authError(error.data) });
             },
             deleteUser: function(resolver, username, callback) {
-                $http.delete(userUrl + "/" + resolver + "/" + username,
+                $http.delete(userUrl + "/" + encodeURIComponent(resolver) + "/" + encodeURIComponent(username),
```
[the remark this answers is not in the document]
Comment (1-am-r00t) [reply privacyidea/static/components/user/factories/user.js]: proposed changes implemented in [bfc9af4](https://github.com/privacyidea/privacyidea/pull/3306/commits/bfc9af4ac9a8d357b4fa1e74452474bf89a60d74)

**Comment by 1-am-r00t** [comment]
> Hi @1-am-r00t thanks for the PR. I think these are the only places where the username is part of the URL.

Hi @plettich,
your requested changes have been implemented in [bfc9af4](https://github.com/privacyidea/privacyidea/pull/3306/commits/bfc9af4ac9a8d357b4fa1e74452474bf89a60d74).

**Review by plettich — APPROVED** [review]

**Pull request merged**